File: math/number theory/primality tests/Roy and Shopping.cpp
problem link-https://www.hackerearth.com/practice/math/number-theory/primality-tests/practice-problems/algorithm/roy-and-shopping-20/?layout=old

#include<bits/stdc++.h>
using namespace std;
int main()
{
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);
    int t,i,j,N=1000000,p[N+1];
    memset(p,0,sizeof(p));
    for(i=2;i<=N;i++)
    if(p[i]==0)
    for(j=i;j<=N;j+=i)
    if(p[j]==0) p[j]=i;

    scanf("%d",&t);
    while(t--)
    {
        int n;
        scanf("%d",&n);
        printf("%d\n",n-p[n]);
    }
    return 0;
}

File: math/number theory/primality tests/Roy and Shopping.java
problem link-https://www.hackerearth.com/practice/math/number-theory/primality-tests/practice-problems/algorithm/roy-and-shopping-20/?layout=old

import java.io.*;

class TestClass {
    public static void main(String args[] ) throws Exception {
        BufferedReader bu=new BufferedReader(new InputStreamReader(System.in));
        int n=Integer.parseInt(bu.readLine()),N=1000000,i,j,p[]=new int[N+1];
        for(i=2;i<=N;i++)
        if(p[i]==0)
        for(j=i;j<=N;j+=i)
        if(p[j]==0) p[j]=i;
        StringBuilder sb=new StringBuilder();
        while(n-->0)
        {
            int k=Integer.parseInt(bu.readLine());
            sb.append(k-p[k]+"\n");
        }
        System.out.print(sb);
    }
}
